File: DataStructures/LinkedLists/other/DoublyLinkedList/GPT/DoublyLinkedList.py
# # 1. 問題分析

# ### 競技プログラミング視点

# * **データ規模**: N, Q ≤ 100 なので、O(NQ) でも十分高速。
# * **実装優先度**: 配列より「双方向リスト」の再現が目的。
# * **速度確保**: Pythonではオブジェクト生成コストがあるため、ループの最適化がポイント。

# ### 業務開発視点

# * **可読性**: `Node` クラス + `DoublyLinkedList` クラスで責務分離。
# * **保守性**: 型ヒントをつけて Pylance で静的解析可能にする。
# * **堅牢性**: 範囲外アクセスを明示的に `ValueError` として扱う。

# ### Python特有の考慮点

# * `list` / `deque` を使えばもっと簡単だが、ここでは「自作双方向リスト」に意味がある。
# * 内包表記での `to_list` 実装など Pythonic に書く。
# * `__slots__` を使うことで `Node` のメモリ効率改善可。

# ---

# # 2. アルゴリズム比較

# | アプローチ         | 時間計算量      | 空間計算量 | 実装コスト | 可読性 | 標準ライブラリ           | 備考            |
# | ------------- | ---------- | ----- | ----- | --- | ----------------- | ------------- |
# | 方法A: deque 利用 | O(1) 挿入/削除 | O(N)  | 低     | 高   | collections.deque | 最速だがリスト自作感なし  |
# | 方法B: heapq利用  | O(n log n) | O(n)  | 中     | 中   | heapq             | 優先度処理向き、今回は不要 |
# | 方法C: 自作双方向リスト | O(n) 挿入/削除 | O(N)  | 高     | 中   | なし                | 本問題の要求に忠実     |

# ---

# # 3. 実装方針

# * **選択したアプローチ**: 方法C（自作双方向リスト）
# * **理由**:

#   * 問題文が「双方向リスト」を明示しているため、構造体で再現。
#   * N, Q が小さいため、線形探索でも十分高速。
#   * 学習的価値（構造理解）も大きい。
# * **最適化ポイント**:

#   * `Node` に `__slots__` を定義し、属性辞書を持たせずメモリ削減。
#   * `insert_at` / `erase_at` は **O(n)** 探索 + **O(1)** 更新。
#   * `to_list` は while ループで単純に走査。

# ---

# # 4. コード実装例

# ```python
from typing import Optional, List


class Node:
    __slots__ = ("value", "prev", "next")

    def __init__(self, value: int):
        self.value: int = value
        self.prev: Optional["Node"] = None
        self.next: Optional["Node"] = None


class DoublyLinkedList:
    def __init__(self) -> None:
        self.head: Optional[Node] = None
        self.tail: Optional[Node] = None
        self.length: int = 0

    def append(self, value: int) -> None:
        """末尾に追加"""
        node = Node(value)
        if self.head is None:
            self.head = self.tail = node
        else:
            assert self.tail is not None
            self.tail.next = node
            node.prev = self.tail
            self.tail = node
        self.length += 1

    def insert_at(self, pos: int, value: int) -> None:
        """1-based位置 pos に挿入"""
        if pos < 1 or pos > self.length + 1:
            raise ValueError("Invalid position for insert")
        if pos == self.length + 1:
            self.append(value)
            return
        node = Node(value)
        if pos == 1:
            assert self.head is not None
            node.next = self.head
            self.head.prev = node
            self.head = node
        else:
            cur = self.head
            for _ in range(pos - 1):
                assert cur is not None
                cur = cur.next
            assert cur is not None
            prev_node = cur.prev
            if prev_node:
                prev_node.next = node
            node.prev = prev_node
            node.next = cur
            cur.prev = node
            if pos == 1:
                self.head = node
        self.length += 1

    def erase_at(self, pos: int) -> None:
        """1-based位置 pos を削除"""
        if pos < 1 or pos > self.length:
            raise ValueError("Invalid position for erase")
        cur = self.head
        for _ in range(pos - 1):
            assert cur is not None
            cur = cur.next
        assert cur is not None
        if cur.prev:
            cur.prev.next = cur.next
        else:
            self.head = cur.next
        if cur.next:
            cur.next.prev = cur.prev
        else:
            self.tail = cur.prev
        self.length -= 1

    def to_list(self) -> List[int]:
        res: List[int] = []
        cur = self.head
        while cur:
            res.append(cur.value)
            cur = cur.next
        return res


def process_list(N: int, Q: int, A: List[int], queries: List[List[int]]) -> List[int]:
    dll = DoublyLinkedList()
    for v in A:
        dll.append(v)
    for q in queries:
        if q[0] == 1:
            _, P, X = q
            dll.insert_at(P, X)
        elif q[0] == 2:
            _, P = q
            dll.erase_at(P)
        else:
            raise ValueError("Invalid query type")
    return dll.to_list()


def main() -> None:
    import sys

    input_data = sys.stdin.read().strip().splitlines()
    N, Q = map(int, input_data[0].split())
    A = [int(x) for x in input_data[1 : 1 + N]]
    queries = [list(map(int, line.split())) for line in input_data[1 + N :]]
    result = process_list(N, Q, A, queries)
    print("\n".join(map(str, result)))


if __name__ == "__main__":
    main()
# ```

# ---

# # 5. 計算量まとめ

# * `insert_at`, `erase_at`: **O(n)**（ノード探索） + **O(1)**（リンク更新）
# * 全体: O(NQ) （N,Q ≤ 100 なので余裕）
# * **Python特有のオーバーヘッド**:

#   * クラス生成コストあり（ただし小規模なので問題なし）
#   * CPythonは属性アクセスが遅め → `__slots__` で軽減

# ---

# # 6. Python固有の最適化観点

# * **CPython最適化**:

#   * `__slots__` により `Node` の辞書オーバーヘッドを削減。
#   * while ループの方が for より速いケースが多い。
# * **メモリ効率化**:

#   * ジェネレータ不要（出力全件を一度に出す必要があるため）。
# * **型ヒント**:

#   * `Optional[Node]`, `List[int]` を明記。
#   * Pylance による補完・検証が効く。

# ---

# ✅ これで **Python 3.8.10 双方向リスト版の完全解答** になります。

File: DataStructures/LinkedLists/other/DoublyLinkedList/GPT/DoublyLinkedList.js
// ## 1. 問題分析

// * **競技プログラミング視点**

//   * N, Q ≤ 100 と小規模。
//   * 直接「双方向リスト」を実装する必要はなく、配列で代用しても十分高速。
//   * 挿入・削除は配列の `splice` を使えば O(n) で処理可能。
//     n ≤ 200 程度なので全く問題なし。

// * **業務開発視点**

//   * 双方向リストを本当に実装する場合、Node クラスを作って `prev` / `next` を管理する。
//   * しかし可読性・保守性の観点では、今回の制約なら配列で表現する方がシンプル。

// * **JavaScript特有の考慮**

//   * V8 は数値配列に最適化されているので、型を統一するのが有利。
//   * `splice` は内部で効率化されており、今回の規模なら高速。

// ---

// ## 2. アルゴリズム比較

// | アプローチ            | 時間計算量       | 空間計算量  | JS実装コスト | 可読性 | 備考      |
// | ---------------- | ----------- | ------ | ------- | --- | ------- |
// | **A: 配列＋splice** | O(NQ) (≦2万) | O(N+Q) | 低       | 高   | 制約内なら最適 |
// | **B: 本当の双方向リスト** | O(1) 挿入/削除  | O(N+Q) | 高       | 中   | 実装コスト大  |
// | **C: 再構築ベース**    | O(NQ)       | O(N+Q) | 中       | 中   | 無駄が多い   |

// ---

// ## 3. 実装方針

// * **選択**: **配列＋splice**
// * **理由**:

//   * N, Q ≤ 100 で十分高速。
//   * 実装が簡潔でエラーバグが少ない。
//   * Node.js 配列操作の内部最適化を活用できる。

// ---

// ## 4. コード実装

// ```javascript
// /**
//  * 双方向リスト操作を配列でシミュレートする
//  * @param {number} N - 初期要素数
//  * @param {number} Q - クエリ数
//  * @param {number[]} A - 初期配列
//  * @param {Array<Array<number>>} queries - クエリ (形式: [1, P, X] または [2, P])
//  * @returns {number[]} 操作後の配列
//  * @throws {Error} 無効なクエリや範囲外アクセス時
//  * @complexity Time: O(N*Q), Space: O(N+Q)
//  */
// function processList(N, Q, A, queries) {
//     if (!Array.isArray(A) || !Array.isArray(queries)) {
//         throw new TypeError("A and queries must be arrays");
//     }

//     /** @type {number[]} */
//     const list = [...A]; // 初期化

//     for (let i = 0; i < Q; i++) {
//         const q = queries[i];
//         if (q[0] === 1) {
//             // INSERT
//             const [, P, X] = q;
//             const M = list.length;
//             if (P < 1 || P > M + 1) {
//                 throw new RangeError("Invalid P in INSERT");
//             }
//             // P は1始まり、spliceは0始まり → P-1
//             list.splice(P - 1, 0, X);
//         } else if (q[0] === 2) {
//             // ERASE
//             const [, P] = q;
//             const M = list.length;
//             if (P < 1 || P > M) {
//                 throw new RangeError("Invalid P in ERASE");
//             }
//             list.splice(P - 1, 1);
//         } else {
//             throw new Error("Invalid query type");
//         }
//     }

//     return list;
// }

// // =======================
// // 実行部分（Node.js標準入力処理）
// // =======================
// function main() {
//     const input = require("fs").readFileSync(0, "utf8").trim().split("\n");
//     const [N, Q] = input[0].split(" ").map(Number);
//     const A = input.slice(1, 1 + N).map(Number);
//     const queries = input.slice(1 + N).map(line => line.split(" ").map(Number));

//     const result = processList(N, Q, A, queries);
//     console.log(result.join("\n"));
// }

// if (require.main === module) {
//     main();
// }
// ```

// ---

// ### ✅ このコードの特徴

// * **配列 `splice`** でシンプルに挿入・削除を実現
// * **入力検証あり** (`RangeError`, `TypeError`)
// * **JSDocコメント付き**
// * **Pure function** (`processList`) と **I/O部分** を分離

// ---

// # 🔹 実装解説

// ## 1. データ構造設計

// ```text
// Node {
//   value: number
//   prev: Node | null
//   next: Node | null
// }

// DoublyLinkedList {
//   head: Node | null
//   tail: Node | null
//   length: number
//   methods:
//     - append(value)        // 末尾に追加
//     - insertAt(pos, value) // P 番目に挿入
//     - eraseAt(pos)         // P 番目を削除
//     - toArray()            // 配列に変換
// }
// ```

// * **append**: 末尾に O(1) で追加可能
// * **insertAt**: 先頭から P 番目を辿って O(n)
// * **eraseAt**: 同上、O(n)
// * **toArray**: head から順に走査 O(n)

// ---

// ## 2. コード実装（Node.js v16.17.1）

// ```javascript
/**
 * 双方向リストのノード
 */
class Node {
  /**
   * @param {number} value
   */
  constructor(value) {
    this.value = value;
    this.prev = null;
    this.next = null;
  }
}

/**
 * 双方向リストの実装
 */
class DoublyLinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
    this.length = 0;
  }

  /**
   * 末尾にノードを追加
   * @param {number} value
   */
  append(value) {
    const node = new Node(value);
    if (!this.head) {
      this.head = this.tail = node;
    } else {
      this.tail.next = node;
      node.prev = this.tail;
      this.tail = node;
    }
    this.length++;
  }

  /**
   * P番目(1-based)に要素を挿入
   * @param {number} pos - 挿入位置 (1 <= pos <= length+1)
   * @param {number} value - 挿入する値
   */
  insertAt(pos, value) {
    if (pos < 1 || pos > this.length + 1) {
      throw new RangeError("Invalid position for insert");
    }
    const node = new Node(value);

    if (pos === 1) {
      // 先頭に挿入
      if (!this.head) {
        this.head = this.tail = node;
      } else {
        node.next = this.head;
        this.head.prev = node;
        this.head = node;
      }
    } else if (pos === this.length + 1) {
      // 末尾に挿入
      this.append(value);
      return;
    } else {
      // 中間に挿入
      let cur = this.head;
      for (let i = 1; i < pos; i++) {
        cur = cur.next;
      }
      const prevNode = cur.prev;
      prevNode.next = node;
      node.prev = prevNode;
      node.next = cur;
      cur.prev = node;
    }
    this.length++;
  }

  /**
   * P番目(1-based)のノードを削除
   * @param {number} pos
   */
  eraseAt(pos) {
    if (pos < 1 || pos > this.length) {
      throw new RangeError("Invalid position for erase");
    }
    let cur = this.head;
    for (let i = 1; i < pos; i++) {
      cur = cur.next;
    }
    if (cur.prev) {
      cur.prev.next = cur.next;
    } else {
      this.head = cur.next; // 先頭削除
    }
    if (cur.next) {
      cur.next.prev = cur.prev;
    } else {
      this.tail = cur.prev; // 末尾削除
    }
    this.length--;
  }

  /**
   * リストを配列に変換
   * @returns {number[]}
   */
  toArray() {
    const res = [];
    let cur = this.head;
    while (cur) {
      res.push(cur.value);
      cur = cur.next;
    }
    return res;
  }
}

/**
 * 双方向リスト操作をシミュレート
 * @param {number} N
 * @param {number} Q
 * @param {number[]} A
 * @param {Array<Array<number>>} queries
 * @returns {number[]}
 */
function processListWithDLL(N, Q, A, queries) {
  const dll = new DoublyLinkedList();
  for (const val of A) {
    dll.append(val);
  }

  for (const q of queries) {
    if (q[0] === 1) {
      const [, P, X] = q;
      dll.insertAt(P, X);
    } else if (q[0] === 2) {
      const [, P] = q;
      dll.eraseAt(P);
    } else {
      throw new Error("Invalid query type");
    }
  }

  return dll.toArray();
}

// =======================
// 実行部分
// =======================
function main() {
  const input = require("fs").readFileSync(0, "utf8").trim().split("\n");
  const [N, Q] = input[0].split(" ").map(Number);
  const A = input.slice(1, 1 + N).map(Number);
  const queries = input.slice(1 + N).map((line) => line.split(" ").map(Number));

  const result = processListWithDLL(N, Q, A, queries);
  console.log(result.join("\n"));
}

if (require.main === module) {
  main();
}
// ```

// ---

// # 🔹 この実装の特徴

// * **配列を使わず、Node構造体で管理**
// * **`insertAt` / `eraseAt` を本当に O(n) 探索 → O(1) 更新で実装**
// * **保守性**: `DoublyLinkedList` クラスに責務を分離
// * **可読性**: 「双方向リストらしい処理」を忠実に再現
